src/objective_types/constraints.jl

```diff
@@ -139,27 +139,13 @@ function OnceDifferentiableConstraints(c!, lx::AbstractVector, ux::AbstractVecto
     xcache = zeros(T, sizex)
     ccache = zeros(T, sizec)
 
-    if is_finitediff(autodiff)
-        ccache2 = similar(ccache)
-        fdtype = finitediff_fdtype(autodiff)
-        jacobian_cache = FiniteDiff.JacobianCache(xcache, ccache,ccache2,fdtype)
-        function jfinite!(J, x)
-            FiniteDiff.finite_difference_jacobian!(J, c!, x, jacobian_cache)
-            J
-        end
-        return OnceDifferentiableConstraints(c!, jfinite!, bounds)
-    elseif is_forwarddiff(autodiff)
-        jac_cfg = ForwardDiff.JacobianConfig(c!, ccache, xcache, chunk)
-        ForwardDiff.checktag(jac_cfg, c!, xcache)
-
-        function jforward!(J, x)
-            ForwardDiff.jacobian!(J, c!, ccache, x, jac_cfg, Val{false}())
-            J
-        end
-        return OnceDifferentiableConstraints(c!, jforward!, bounds)
-    else
-        error("The autodiff value $autodiff is not support. Use :finite or :forward.")
+    backend = get_adtype(autodiff, chunk)
+    jac_prep = DI.prepare_jacobian(c!, ccache, backend, xcache)
+    function j!(_j, _x)
+        DI.jacobian!(c!, ccache, _j, jac_prep, backend, _x)
```

Comment:
Maybe these could be made callable structs to avoid closing over `c!`, `ccache`, `jac_prep` and `backend`?

Reply:
These were already closures in the existing code, I thought I'd go for the "minimum diff" changes and then we could improve later on

Reply:
Yeah, I saw that. It seemed you're closing over more variables in the DI version though, so it might be even more valuable to change the design.

Reply:
I seem to recall that closures are not an issue if the variable we close over does not get assigned to more than once? So we might be good here?
In any case, to hunt down this kind of type inference barriers we would also need to improve annotation of functions in the various structs of the package.

Reply:
I don't know. Even if it's currently the case, personally I wouldn't rely on it since in my experience the exact behaviour of the compiler is inherently unstable.

Reply:
See for instance the explanations in https://discourse.julialang.org/t/can-someone-explain-closures-to-me/105605